### ISSUE TITLE
ODC suppression not taken into account in ODT

### ISSUE THREAD

**Issue by dldnhf** [DependencyTrack/dependency-track#310]
### Issue Type:
  - [x] defect report
  - [ ] enhancement request
  
  
### Current Behavior:
When uploading an ODC XML report, a suppression from the suppressions.xml is ignored by ODT, even though it correctly shows up in the dependency-check-report.xml produced by ODT ("suppressedIdentifier" and "suppressedVulnerability" - see attached [copy](https://github.com/DependencyTrack/dependency-track/files/3033851/dependency-check-report-odt.xml.txt) of
~\.dependency-track\dependency-check\reports\dependency-check-report.xml).

The suppressed vulnerability is still listed in ODT, and not suppressed.

![odt](https://user-images.githubusercontent.com/247258/55401964-89daae00-5552-11e9-8ead-e22a7390445d.jpg)

### Steps to Reproduce (if defect):
 - Create project in ODC
 - Provide suppression in ~\.dependency-track\dependency-check\suppressions.xml (see [suppressions.xml](https://github.com/DependencyTrack/dependency-track/files/3033849/suppressions.xml.txt))
 - PUT api/v1/scan with ODC XML (see [dependency-check-report.xml](https://github.com/DependencyTrack/dependency-track/files/3033850/dependency-check-report.xml.txt)) in payload 

### Expected Behavior:

Component suppressed by ODC suppression file must not have any vulnerabilities.

### Environment:
  
  - Dependency-Track Version: 3.4.0
  - Distribution: Executable WAR
  - BOM Format & Version: Dependency Check 4.0.2 XML
  - Database Server: H2

  - Browser: Chrome

**Comment by stevespringett** [comment]
For clarification, ODT is not a storage location for ODC reports.

- ODC does not have the necessary data and data integrity in place to honor suppressions in ODT. ODC is currently missing:
  - Who perform the analysis
  - Analysis state changes (exploitable, not an issue, etc)
  - Scope (could be a false positive or could be a case where the finding isn't relevant to the project). So global or project scope.
  - Timestamp for when every event occurred
  - The above being signed and integrity checked
- ODC is usually not the only scanner enabled in ODT, so suppressing a finding in ODC would have no affect on other scanners that ODT incorporates
- ODT requires BOMs and ODC has a pseudo-bom, but ODT treats it as a BOM
- Once BOMs are imported, a scan is run. So whatever suppressions were in the ODC report uploaded would have no affect on future scans

See: https://docs.dependencytrack.org/best-practices/

The REST API in Dependency-Track allows you to perform project and global suppressions. So you could write something that reads in the suppressions and then suppresses them in Dependency-Track.

**Comment by dldnhf** [comment]
@stevespringett thanks a lot for your clarification. I was aware of most of that, and think there's a misunderstanding about my concrete issue:

I didn't expect ODT to honor any suppressions in the uploaded ODC report itself, but I provided a suppressions.xml and expected ODT to take it into account when running the scan. If this is a wrong expectation, I'm clearly missing something - or what's the point in the possibility to provide a suppression file then? 

I get it that it's not best practice, but I only want to use these global ODC suppressions for flat-out wrong dependencies like in my example - _org.eclipse.equinox.app_ simply has nothing to do with _cpe:/a:app_project:app_.

Note that the _dependency-check-report.xml_ I mentioned, with the "_suppressedIdentifier_" and "_suppressedVulnerability_" elements, was produced by the ODT scan (i.e., this was NOT the report I had uploaded). I only mentioned it to help narrow down the issue. So why is this vulnerability not suppressed in ODT? If a different scanner (other than ODC) is responsible for it, how can I find out?

**Comment by stevespringett** [comment]
Thanks for clarifying. I missed the fact you were using suppressions in dependency-track itself and assumed you were using them when the dependency-check scan was originally run that produced the xml. So most of what I said still applies, but not all.

However, I looked at the suppression and there a few things wrong and a few assumptions being made. Dependency-Check includes many file-type analyzers, things that look at poms, jar files, etc to extract evidence. Dependency-Track only uses the best evidence (high or highest confidence on a few fields) to determine facts. When Dependency-Track performs a Dependency-Check scan, it supplies these facts to Dependency-Check and in doing so Dependency-Check does not run its file-type analyzers because there are no files. So it uses the facts Dependency-Track supplied. It's these facts that needs to be used in the authoring of suppressions. Things like name, group, version, hashes, etc that Dependency-Track knows about should be used. 

With that said, the suppression file format schema incorrectly assumes that all dependency evidence will be derived from file-type analyzers. So instead of abstracting it out (like group, name, version, etc), the suppressions rely on things that are irrelevant in cases where file-type analyzers are not executed. 'gav' for example will never suppress findings in Dependency-Track since gav is a Maven construct (file-type analyzer) and that evidence will never be supplied by Dependency-Track.  It's also one of the reasons why the base suppressions included with Dependency-Check do not work in Dependency-Track. So you actually end up with more noise than normal. 

Inspecting suppressions.xml, I think the only elements that could be used are:

- sha1

and a choice of zero or more of the following:
- cpe 
- cve
- cwe

Because Dependency-Track supplies facts to Dependency-Check to use as evidence, I'm a little unsure what the values for CPE would be. So you may have to play with that to get it to work. You'll likely have better luck using a combination of sha1 and cve.

**Comment by stevespringett** [comment]
One note, Dependency-Check v5 may provide a workaround for some of the above limitations. Support for v5 may or may not end up in the next version of Dependency-Track. There are a lot of changes, it's untested, and will likely need some time to stabilize. However, when it's eventually incorporated into Dependency-Track, Check now uses PackageURL and there's built-in functionality to take PackageURL evidence and transpose that into gav and other syntax that may help in various suppressions.

**Comment by dldnhf** [comment]
Thanks, I appreciate your insights!

I've tried sha1+cpe as well as sha1+cve, both work up to the OD**T**-internally created XML report (where the suppressions show up fine), but both don't have any effect in the ODT model. 

I've debugged `DependencyCheckTask.processResults()`, where this internal XML report is read: `vulnerabilities.getVulnerabilities()` returns `null` (only `suppressedVulnerabilities` is there), hence the vulnerability is not added here... `QueryManager.addVulnerability()` wasn't entered from anywhere else either. Any hints how the vulnerability ends up in the component then? 

As of now, I've never seen any ODC suppressions work in ODT. Does anyone have an example where it does work?

**Comment by stevespringett** [comment]
Dependency-Check suppressions will not suppress vulnerabilities that already exist in the database.

Once a vulnerability is identified on a component, it has to go through the Dependency-Track auditing process. Orgs will typically have notifications or webhooks that trigger specific events outside of Dependency-Track. So if a new vulnerability comes in, a JIRA ticket may be automatically created or pull request may already have taken place. The Dependency-Track auditing process also produces events. So marking things as false positives may close those JIRA tickets for example. Blindly suppressing the vulnerability without an audit trail (via dependency-check suppressions) would invalidate this workflow.

Support for the global dependency-check suppression file is intended for a single purpose only; to provide a way for orgs that have already developed their own org-specific suppression file, a way to utilize that as a base on a new Dependency-Track installation. However, moving forward, use of the Dependency-Track auditing process is the only supported way to suppress findings once they're in the database.

**Comment by dldnhf** [comment]
I did delete the component as well as the project before every retest - I would've expected that the component and the association to the vulnerability is gone then!? I've now deleted the .dependency-track directory (including the H2 files) and started from scratch - and _now_ the suppression is in effect. 

Also, our use case for the ODC suppressions is exactly the one you described. We're looking into generating BOMs to use it for ODT, but given our codebase and the fact that we have plenty of non-maven dependencies as well, this might not be trivial. Therefore, we want to use what we already have, but still leverage ODT for process integrations. As far as I've understood, ODT assumes that BOMs are always correct, hence there is no way to suppress a component itself (unlike with ODC, where due to the heuristic approach false positives are common, and it's easy to suppress a false dependency once and for all).
